stdlib/asm/crypto/dsa/rpo_falcon512.masm

```diff
@@ -125,275 +125,185 @@ end
-#! the advice stack and lays it right after `s2`.
+#! After that, it loads the claimed polynomial `h * s2` in Z_Q[x] where Q is the Miden VM prime
+#! from the advice stack and lays it right after `s2`.
 #! The hash of `h`, `s2` and the claimed product is also computed and the first two field elements
-#! of the digest (i.e., the Fiat-Shamir challenge) are returned on the stack alongside
-#! the incremented pointer.
-#!
-#! Input: [ptr, PK, ...]
-#! Output: [tau1, tau0, ptr + 512*4 ...]
-#!
-#! Cycles: 5049
-export.load_h_s2_and_product.4
+#! of the digest (i.e., the Fiat-Shamir challenge) are used in order to check that 
+#! pi == h * s2 in Z_Q[x] by evaluating both sides at the random point. Note that since we need
+#! to hash the polynomials before computing their evaluations, the procedure receives
+#! the evaluation point non-deterministically so that it can perform the evaluations while hashing
+#! is still going on. After hashing of the polynomials is finished, the procedure checks that
+#! the evaluation point was indeed derived correctly from the hash of all 3 polynomials.
+#!
+#! Inputs:
+#!   Operand stack: [ptr, PK, ...]
+#!   Advice stack:  [tau0, tau1, h_0, ..., h_511, s2_0, ..., s2_511, pi_0, ..., pi_1022, ...]
+#! Outputs:
+#!   Operand stack: []
+#!   Advice stack:  []
+#!
+#! Cycles: 4430
+export.load_h_s2_and_product.12
+    # We will store in local addresses:
+    # 1. 0..4 the public key,
+    # 2. 4..8 the inverse of the evaluation point tau as [tau_inv0, tau_inv1, 0, 0],
+    # 3. 8..12 the evaluation point tau as [tau0, tau1, 0, 0],
+    
+    # 1) Set up the stack for loading the coefficients of the polynomials, evaluating and hashing them
+    
+    ## a) Set up the accumulator for `horner_eval_base` and the memory pointers
+    push.0.0
+    locaddr.4
+    movup.3
 
-    # 1) Store PK for later comparison
-    movdn.4
+    ## b) Save PK to later compare it with the hash of the h polynomial
+    swapw
     loc_storew.0
+    # => [PK, ptr, alpha_inv, acc1, acc0, ...]
 
-    # 2) Prepare stack and load h polynomial. We also range check the coefficients of h.
+    ## c) Prepare the capacity portion of the state of the hasher
     padw swapw
-    padw
+    # => [Y, 0, 0, 0, 0, ptr, alpha_inv, acc1, acc0, ...], where Y is a "garbage" word
+
+    ## d) Load the evaluation point tau from the advice tape, compute its inverse and save both
+    ## Note that we will evaluate the polynomials at `tau_inv = tau^{-1}` as we are loading the coefficients
+    ## in the normal order i.e., not in reversed order required for Horner evaluation.
+    ## This means that the equality we will be checking is 
+    ## pi(tau_inv) * tau^1023 == h(tau_inv) * tau^511 * s2(tau_inv) * tau^511
+    ##
+    ## which simplifies to
+    ##
+    ## pi(tau_inv) * tau == h(tau_inv) * s2(tau_inv)
+    adv_push.2 
+    dup.1 dup.1 ext2inv
+    push.0.0
+    loc_storew.4
```

Comment:
Could we add stack state after this instruction?

Reply:
done